README.md
# ACF - Advanced Taxonomy Selector

A field for Advanced Custom Fields which allows you to create a field where users can select terms or taxonommies flexibly.

## Description

A more advanced taxonomy selector for ACF. You can create a taxonomy or a term selector and restrict each based on post type or taxonomy. You can also allow for null and determine the return format.

#### Thanks
- [Advanced Custom Fields](http://www.advancedcustomfields.com/) for the awesome base plugin.
- [Muhamad Ulum](https://www.iconfinder.com/oelhoem) for the T icon.

#### Translations

The plugin is currently available in English and Hungarian. Please feel free to submit any new languages via a pull request, I'd be mighty thankful.

#### Useful Links

This Github repository is for the development of this plugin. If you would like to read installation and in-depth usage instructions you might want to look at the WordPress plugin page instead.

- [Plugin Page](https://wordpress.org/plugins/acf-advanced-taxonomy-selector/)
- [SVN Repository](http://plugins.svn.wordpress.org/acf-advanced-taxonomy-selector/)
- [ACF Plugin](https://wordpress.org/plugins/advanced-custom-fields/)
- [ACF Home Page](http://www.advancedcustomfields.com/)
- [ACF Documentation](http://www.advancedcustomfields.com/resources/)
- [ACF Field Template](https://github.com/elliotcondon/acf-field-type-template)
- [ACF on Github](https://github.com/elliotcondon/acf)

# Want To Help?

If you like the plugin and you like helping others out there are a few things you can do:

- **[Buy ACF Pro](http://www.advancedcustomfields.com/pro/)**
- **[Review the plugin](https://wordpress.org/support/view/plugin-reviews/acf-advanced-taxonomy-selector)**
- **Submit a translation** If you speak another language goodly, you can submit a language file, I'd be mighty thankful! Take a look at the lang directory to see what languages we already have. If a language isn't there create one and submit a pull request. If you have no idea what I'm talking about drop me a line and I'll help you out
- **[Tip me on Gratipay](https://gratipay.com/danielpataki/)**


CHANGELOG.md
### 3.0.0 (2015-04-21)
- Updated for WordPress 4.2
- Added Hungarian translation
- Reworked structure

### 2.0
- Dropped ACF 3 support
- Added ACF 5 support

### 1.0
- Initial Release.


advanced_taxonomy_selector-v3.php
<?php

class acf_field_advanced_taxonomy_selector extends acf_Field
{

	// vars
	var $settings, // will hold info such as dir / path
		$defaults; // will hold default field options


	/*--------------------------------------------------------------------------------------
	*
	*	Constructor
	*	- This function is called when the field class is initalized on each page.
	*	- Here you can add filters / actions and setup any other functionality for your field
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/

	function __construct($parent)
	{

		// do not delete!
  	parent::__construct($parent);

		// vars
		$this->name = 'advanced_taxonomy_selector';
		$this->label = __('Advanced Taxonomy Selector');
		$this->category = __("Relational",'acf'); // Basic, Content, Choice, etc
		$this->defaults = array(
			'taxonomies' => '',
			'type' => 'terms',
			'field_type' => 'multiselect',
			'return_value' => 'term_id'
		);



    // settings
		$this->settings = array(
			'path' => apply_filters('acf/helpers/get_path', __FILE__),
			'dir' => apply_filters('acf/helpers/get_dir', __FILE__),
			'version' => '1.0.0'
		);

  }


 	/*
  *  helpers_get_path
  *
  *  @description: calculates the path (works for plugin / theme folders)
  *  @since: 3.6
  *  @created: 30/01/13
  */

  function helpers_get_path($file)
  {
    return trailingslashit(dirname($file));
  }


  /*
  *  helpers_get_dir
  *
  *  @description: calculates the directory (works for plugin / theme folders)
  *  @since: 3.6
  *  @created: 30/01/13
  */

  function helpers_get_dir($file)
  {
    $dir = trailingslashit(dirname($file));
    $count = 0;


    // sanitize for Win32 installs
    $dir = str_replace('\\', '/', $dir);


    // if file is in plugins folder
    $wp_plugin_dir = str_replace('\\', '/', WP_PLUGIN_DIR);
    $dir = str_replace($wp_plugin_dir, WP_PLUGIN_URL, $dir, $count);


    if($count < 1)
    {
      // if file is in wp-content folder
      $wp_content_dir = str_replace('\\', '/', WP_CONTENT_DIR);
      $dir = str_replace($wp_content_dir, WP_CONTENT_URL, $dir, $count);
    }


    if($count < 1)
    {
      // if file is in ??? folder
      $wp_dir = str_replace('\\', '/', ABSPATH);
      $dir = str_replace($wp_dir, site_url('/'), $dir);
    }

    return $dir;
  }


	/*--------------------------------------------------------------------------------------
	*
	*	create_options
	*	- this function is called from core/field_meta_box.php to create extra options
	*	for your field
	*
	*	@params
	*	- $key (int) - the $_POST obejct key required to save the options to the field
	*	- $field (array) - the field object
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/

	function create_options($key, $field)
	{
		$field = array_merge($this->defaults, $field);

		// key is needed in the field names to correctly save the data
		$key = $field['name'];

		$taxonomies = get_taxonomies( array( '_builtin' => false, 'public' => true ), 'objects' );
		$choices = array( 'all' => __( 'All Taxonomies', 'acf' ) );
		foreach ( $taxonomies as $slug => $taxonomy ) {
			$choices[$slug] = $taxonomy->label;
		}

		?>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Type", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'radio',
			'name'    =>  'fields[' . $key . '][type]',
			'value'   =>  $field['type'],
			'choices' =>  array(
				'terms' => __( 'Choose Terms', 'acf' ),
				'taxonomy'  => __( 'Choose Taxonomies', 'acf' ),
			)
		));

		?>
	</td>
</tr>
		
<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Taxonomies", 'acf'); ?></label>
		<p class="description"><?php _e("Set taxonomies users will be able to choose from", 'acf'); ?></p>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'select',
			'name'    =>  'fields[' . $key . '][taxonomies]',
			'value'   =>  $field['taxonomies'],
			'multiple'  =>  true,
			'choices' =>  $choices
		));

		?>
	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Field Type", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'select',
			'name'    =>  'fields[' . $key . '][field_type]',
			'value'   =>  $field['field_type'],
			'choices' =>  array(
				'multiselect'  => __( 'Multiselect', 'acf' ),
				'select'       => __( 'Select', 'acf' )
			)
		));

		?>
	</td>
</tr>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Allow Null?", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'radio',
			'name'    =>  'fields[' . $key . '][allow_null]',
			'layout'  => 'horizontal',
			'value'   =>  $field['allow_null'],
			'choices' =>  array(
				1 => __( 'Yes', 'acf' ),
				0  => __( 'No', 'acf' ),
			)
		));

		?>
	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Return Value", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'radio',
			'name'    =>  'fields[' . $key . '][return_value]',
			'value'   =>  $field['return_value'],
			'choices' =>  array(
				'term_id' => __( 'Term ID', 'acf' ),
				'object'  => __( 'Object', 'acf' ),
			)
		));

		?>
	</td>
</tr>



		<?php

	}




	/*--------------------------------------------------------------------------------------
	*
	*	create_field
	*	- this function is called on edit screens to produce the html for this field
	*
	*	@author Elliot Condon
	*	@since 2.2.0
	*
	*-------------------------------------------------------------------------------------*/
	function create_field( $field )
	{

		$taxonomies = $field['taxonomies'];
		if( empty( $field['taxonomies'] ) || ( !empty( $field['taxonomies'] ) && in_array( 'all', $field['taxonomies'] ) ) ) {
			$taxonomies = get_taxonomies( array( '_builtin' => false, 'public' => true ) );
		}

		$taxonomies = array_values( $taxonomies );
		
		if ( $field['data_type'] == 'terms' ) {
			$term_list = array();

			foreach ( $taxonomies as $taxonomy_slug ) {
				$taxonomy = get_taxonomy( $taxonomy_slug );
				$terms = get_terms( $taxonomy_slug );
				$term_list[$taxonomy_slug] = array(
					'name' => $taxonomy->label,
					'slug' => $taxonomy_slug,
					'terms' => array()
				);
				foreach ( $terms as $term ) {
					$term_list[$taxonomy_slug]['terms'][] = array(
						'name' => $term->name,
						'slug' => $term->slug,
						'term_id' => $term->term_id
					);
				}
			}

			if ( $field['field_type'] == 'multiselect' || $field['field_type'] == 'select' ) :
			?>
			<div>
				<?php if ( $field['field_type'] == 'multiselect' ) : ?>
				<select name="<?php echo $field['name'] ?>[]" multiple='multiple'>
				<?php else : ?>
				<select name="<?php echo $field['name'] ?>[]">
				<?php endif ?>
					<?php if( !empty( $field['allow_null'] ) ) : ?>
						<option value=''><?php _e( 'All Taxonomies', 'acf' ) ?></option>
					<?php endif ?>				
					<?php foreach ( $term_list as $taxonomy ) : ?>
					<optgroup label='<?php echo $taxonomy['name'] ?>'>
						<?php
							foreach( $taxonomy['terms'] as $term ) :
							$value = $taxonomy['slug'] . '_' . $term['term_id'];
							$selected = ( in_array( $value, $field['value'] ) ) ? 'selected="selected"' : '';
						?>
							<option <?php echo $selected ?> value='<?php echo $value ?>'><?php echo $term['name'] ?></option>
						<?php endforeach ?>
					</optgroup>
					<?php endforeach ?>
				</select>
			</div>
			<?php

			endif;

		}
		else {

			if ( $field['field_type'] == 'multiselect' || $field['field_type'] == 'select' ) :
			?>
			<div>
				<?php if ( $field['field_type'] == 'multiselect' ) : ?>
				<select name="<?php echo $field['name'] ?>[]" multiple='multiple'>
				<?php else : ?>
				<select name="<?php echo $field['name'] ?>[]">
				<?php endif ?>
					<?php 
						foreach ( $taxonomies as $taxonomy_slug ) : 
						$taxonomy = get_taxonomy( $taxonomy_slug );
						$selected = ( 
							( is_array( $field['value'] ) && in_array( $taxonomy_slug, $field['value'] ) )
							||
							( is_string( $field['value'] ) && $taxonomy_slug == $field['value'] )
							||
							( empty( $field['value'] ) && $taxonomy_slug == 'bsd_item_type' )
						) ? 'selected="selected"' : ''
					?>
						<option <?php echo $selected ?> value='<?php echo $taxonomy_slug ?>'><?php echo $taxonomy->label ?></option>
					<?php endforeach ?>
				</select>
			</div>
			<?php

			endif;


		}


	}



}



}

?>


advanced_taxonomy_selector-v4.php
<?php

class acf_field_advanced_taxonomy_selector extends acf_field
{
	// vars
	var $settings, // will hold info such as dir / path
		$defaults; // will hold default field options


	/*
	*  __construct
	*
	*  Set name / label needed for actions / filters
	*
	*  @since	3.6
	*  @date	23/01/13
	*/

	function __construct()
	{
		// vars
		$this->name = 'advanced_taxonomy_selector';
		$this->label = __('Advanced Taxonomy Selector');
		$this->category = __("Relational",'acf'); // Basic, Content, Choice, etc
		$this->defaults = array(
			'taxonomies' => '',
			'data_type' => 'terms',
			'field_type' => 'multiselect',
			'return_value' => 'term_id'
		);

		// do not delete!
    parent::__construct();


    // settings
		$this->settings = array(
			'path' => apply_filters('acf/helpers/get_path', __FILE__),
			'dir' => apply_filters('acf/helpers/get_dir', __FILE__),
			'version' => '1.0.0'
		);

	}


	/*
	*  create_options()
	*
	*  Create extra options for your field. This is rendered when editing a field.
	*  The value of $field['name'] can be used (like bellow) to save extra data to the $field
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field	- an array holding all the field's data
	*/

	function create_options($field)
	{
		$field = array_merge($this->defaults, $field);

		// key is needed in the field names to correctly save the data
		$key = $field['name'];

		$taxonomies = get_taxonomies( array( '_builtin' => false, 'public' => true ), 'objects' );
		$choices = array( 'all' => __( 'All Taxonomies', 'acf' ) );
		foreach ( $taxonomies as $slug => $taxonomy ) {
			$choices[$slug] = $taxonomy->label;
		}

		?>
<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Type", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'radio',
			'name'    =>  'fields[' . $key . '][data_type]',
			'value'   =>  $field['data_type'],
			'choices' =>  array(
				'terms' => __( 'Choose Terms', 'acf' ),
				'taxonomy'  => __( 'Choose Taxonomies', 'acf' ),
			)
		));

		?>
	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Taxonomies", 'acf'); ?></label>
		<p class="description"><?php _e("Set taxonomies users will be able to choose from", 'acf'); ?></p>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'select',
			'name'    =>  'fields[' . $key . '][taxonomies]',
			'value'   =>  $field['taxonomies'],
			'multiple'  =>  true,
			'choices' =>  $choices
		));

		?>
	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Field Type", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'select',
			'name'    =>  'fields[' . $key . '][field_type]',
			'value'   =>  $field['field_type'],
			'choices' =>  array(
				'multiselect'  => __( 'Multiselect', 'acf' ),
				'select'       => __( 'Select', 'acf' )
			)
		));

		?>
	</td>
</tr>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Return Value", 'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'    =>  'radio',
			'name'    =>  'fields[' . $key . '][return_value]',
			'value'   =>  $field['return_value'],
			'choices' =>  array(
				'term_id' => __( 'Term ID / Taxonomy Slug', 'acf' ),
				'object'  => __( 'Term Object / Taxonomy Object', 'acf' ),
			)
		));

		?>
	</td>
</tr>



		<?php

	}


	/*
	*  create_field()
	*
	*  Create the HTML interface for your field
	*
	*  @param	$field - an array holding all the field's data
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*/

	function create_field( $field )
	{

		$taxonomies = $field['taxonomies'];
		if( empty( $field['taxonomies'] ) || ( !empty( $field['taxonomies'] ) && in_array( 'all', $field['taxonomies'] ) ) ) {
			$taxonomies = get_taxonomies( array( '_builtin' => false, 'public' => true ) );
		}

		$taxonomies = array_values( $taxonomies );
		
		if ( $field['data_type'] == 'terms' ) {
			$term_list = array();

			foreach ( $taxonomies as $taxonomy_slug ) {
				$taxonomy = get_taxonomy( $taxonomy_slug );
				$terms = get_terms( $taxonomy_slug );
				$term_list[$taxonomy_slug] = array(
					'name' => $taxonomy->label,
					'slug' => $taxonomy_slug,
					'terms' => array()
				);
				foreach ( $terms as $term ) {
					$term_list[$taxonomy_slug]['terms'][] = array(
						'name' => $term->name,
						'slug' => $term->slug,
						'term_id' => $term->term_id
					);
				}
			}

			if ( $field['field_type'] == 'multiselect' || $field['field_type'] == 'select' ) :
			?>
			<div>
				<?php if ( $field['field_type'] == 'multiselect' ) : ?>
				<select name="<?php echo $field['name'] ?>[]" multiple='multiple'>
				<?php else : ?>
				<select name="<?php echo $field['name'] ?>[]">
				<?php endif ?>
					<?php foreach ( $term_list as $taxonomy ) : ?>
					<optgroup label='<?php echo $taxonomy['name'] ?>'>
						<?php
							foreach( $taxonomy['terms'] as $term ) :
							$value = $taxonomy['slug'] . '_' . $term['term_id'];
							$selected = ( in_array( $value, $field['value'] ) ) ? 'selected="selected"' : '';
						?>
							<option <?php echo $selected ?> value='<?php echo $value ?>'><?php echo $term['name'] ?></option>
						<?php endforeach ?>
					</optgroup>
					<?php endforeach ?>
				</select>
			</div>
			<?php

			endif;

		}
		else {

			if ( $field['field_type'] == 'multiselect' || $field['field_type'] == 'select' ) :
			?>
			<div>
				<?php if ( $field['field_type'] == 'multiselect' ) : ?>
				<select name="<?php echo $field['name'] ?>[]" multiple='multiple'>
				<?php else : ?>
				<select name="<?php echo $field['name'] ?>[]">
				<?php endif ?>
					<?php 
						foreach ( $taxonomies as $taxonomy_slug ) : 
						$taxonomy = get_taxonomy( $taxonomy_slug );
						$selected = ( 
							( is_array( $field['value'] ) && in_array( $taxonomy_slug, $field['value'] ) )
							||
							( is_string( $field['value'] ) && $taxonomy_slug == $field['value'] )
							||
							( empty( $field['value'] ) && $taxonomy_slug == 'bsd_item_type' )
						) ? 'selected="selected"' : ''
					?>
						<option <?php echo $selected ?> value='<?php echo $taxonomy_slug ?>'><?php echo $taxonomy->label ?></option>
					<?php endforeach ?>
				</select>
			</div>
			<?php

			endif;


		}


	}



}


// create field
new acf_field_advanced_taxonomy_selector();

?>


acf-advanced_taxonomy_selector-v4.php
<?php

class acf_field_advanced_taxonomy_selector extends acf_field {

	// vars
	var $settings, // will hold info such as dir / path
		$defaults; // will hold default field options


	/*
	*  __construct
	*
	*  Set name / label needed for actions / filters
	*
	*  @since	3.6
	*  @date	23/01/13
	*/

	function __construct()
	{
		// vars
		$this->name = 'advanced_taxonomy_selector';
		$this->label = __('Advanced Taxonomy Selector');
		$this->category = __("Relational",'acf'); // Basic, Content, Choice, etc
		$this->defaults = array(
			'taxonomies' => '',
			'data_type' => 'terms',
			'field_type' => 'multiselect',
			'allow_null' => true,
			'return_value' => 'term_id'
		);

	   	parent::__construct();

	}


	/*
	*  create_options()
	*
	*  Create extra options for your field. This is rendered when editing a field.
	*  The value of $field['name'] can be used (like below) to save extra data to the $field
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$field	- an array holding all the field's data
	*/

	function create_options( $field )
	{
		// defaults?
		/*
		$field = array_merge($this->defaults, $field);
		*/

		// key is needed in the field names to correctly save the data
		$key = $field['name'];


		// Create Field Options HTML
		?>
<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Type",'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'radio',
			'name'		=>	'fields['.$key.'][data_type]',
			'value'		=>	$field['data_type'],
			'choices' =>  array(
				'terms' => __( 'Choose Terms', 'acf' ),
				'taxonomy'  => __( 'Choose Taxonomies', 'acf' ),
			)
		));

		?>
	</td>
</tr>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Taxonomies",'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'select',
			'name'		=>	'fields['.$key.'][taxonomies]',
			'value'		=>	$field['taxonomies'],
			'multiple'      =>  true,
			'choices'       =>  $this->taxonomies_array()
		));

		?>
	</td>
</tr>


<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Field Type",'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'select',
			'name'		=>	'fields['.$key.'][field_type]',
			'value'		=>	$field['field_type'],
			'choices' =>  array(
				'multiselect'  => __( 'Multiselect', 'acf' ),
				'select'       => __( 'Select', 'acf' )
			)
		));

		?>
	</td>
</tr>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Allow Null?",'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'radio',
			'name'		=>	'fields['.$key.'][allow_null]',
			'value'		=>	$field['allow_null'],
			'layout'  => 'horizontal',
			'choices' =>  array(
				1 => __( 'Yes', 'acf' ),
				0  => __( 'No', 'acf' ),
			)
		));

		?>
	</td>
</tr>

<tr class="field_option field_option_<?php echo $this->name; ?>">
	<td class="label">
		<label><?php _e("Return Value",'acf'); ?></label>
	</td>
	<td>
		<?php

		do_action('acf/create_field', array(
			'type'		=>	'radio',
			'name'		=>	'fields['.$key.'][return_value]',
			'value'		=>	$field['return_value'],
			'choices' =>  array(
				'term_id' => __( 'Term ID / Taxonomy Slug', 'acf' ),
				'object'  => __( 'Term Object / Taxonomy Object', 'acf' ),
			)
		));

		?>
	</td>
</tr>


		<?php

	}



	/*
	*  create_field()
	*
	*  Create the HTML interface for your field
	*
	*  @param	$field - an array holding all the field's data
	*
	*  @type	action
	*  @since	3.6
	*  @date	23/01/13
	*/

	function create_field( $field )
	{
		call_user_func( array( $this, 'render_field_' . $field['data_type'] ), $field );
	}



	/*
	*  Render Terms Field
	*
	*  Displays the field for selecting terms
	*
	*  @param array $field
	*
	*/

	function render_field_terms( $field ) {
		$taxonomies = $this->get_taxonomies_from_selection( $field );
		$multiple = ( $field['field_type'] == 'multiselect' ) ? 'multiple="multiple"' : '';

		foreach( $taxonomies as $slug => $taxonomy ) {
			if( wp_count_terms( $slug ) == 0 ) {
				unset( $taxonomies[$slug] );
			}
		}
		?>
		<select name="<?php echo $field['name'] ?>[]" <?php echo $multiple ?>>
			<?php if( $field['allow_null'] == true ) : ?>
			<option value=''><?php _e( 'None', 'acf-advanced_taxonomy_selector' ) ?></option>
			<?php endif ?>
			<?php
				foreach( $taxonomies as $taxonomy ) :
			?>
				<optgroup label='<?php echo $taxonomy->label ?>'>
					<?php
						$terms = get_terms( $taxonomy->name, array( 'hide_empty' => false ) );
						foreach( $terms as $term ) :
							$selected = ( !empty( $field['value'] ) && in_array( $taxonomy->name . '_' . $term->term_id, $field['value'] ) ) ? 'selected="selected"' : '';
					?>
						<option <?php echo $selected ?> value='<?php echo $taxonomy->name ?>_<?php echo $term->term_id ?>'><?php echo $term->name ?></option>
					<?php endforeach ?>
				</optgroup>
			<?php endforeach ?>
		</select>

		<?php
	}

	/*
	*  Render Taxonomy Field
	*
	*  Displays the field for selecting taxonomies
	*
	*  @param array $field
	*
	*/
	function render_field_taxonomy( $field ) {
		$taxonomies = $this->get_taxonomies_from_selection( $field );
		$multiple = ( $field['field_type'] == 'multiselect' ) ? 'multiple="multiple"' : '';
		?>

		<select name="<?php echo $field['name'] ?>[]" <?php echo $multiple ?>>
			<?php if( $field['allow_null'] == true ) : ?>
			<option value=''><?php _e( 'None', 'acf-advanced_taxonomy_selector' ) ?></option>
			<?php endif ?>

			<?php
				foreach( $taxonomies as $taxonomy ) :
					$selected = ( !empty( $field['value'] ) && in_array( $taxonomy->name, $field['value'] ) ) ? 'selected="selected"' : '';
			?>
				<option <?php echo $selected ?> value='<?php echo $taxonomy->name ?>'><?php echo $taxonomy->label ?></option>
			<?php endforeach ?>
		</select>

		<?php
	}

	/*
	*  update_value()
	*
	*  This filter is applied to the $value before it is saved in the db
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$value (mixed) the value found in the database
	*  @param	$post_id (mixed) the $post_id from which the value was loaded
	*  @param	$field (array) the field array holding all the field options
	*  @return	$value
	*/

	function update_value( $value, $post_id, $field ) {

		if( $value == array( 0 => '' ) ) {
			return '';
		}

		return $value;

	}


	/*
	*  format_value()
	*
	*  This filter is appied to the $value after it is loaded from the db and before it is returned to the template
	*
	*  @type	filter
	*  @since	3.6
	*  @date	23/01/13
	*
	*  @param	$value (mixed) the value which was loaded from the database
	*  @param	$post_id (mixed) the $post_id from which the value was loaded
	*  @param	$field (array) the field array holding all the field options
	*
	*  @return	$value (mixed) the modified value
	*/


	function format_value( $value, $post_id, $field ) {

		if( empty($value) ) {
			return $value;
		}

		if( $field['data_type'] == 'terms' ) {
			foreach( $value as $i => $val ) {
				$term = substr( $val, strrpos( $val, '_' ) + 1 );
				if( $field['return_value'] == 'object' ) {
					$taxonomy = substr( $val, 0, strrpos( $val, '_' ) );
					$term = get_term( $term, $taxonomy );
				}
				$value[$i] = $term;
			}
		}
		elseif( $field['data_type'] == 'taxonomy' && $field['return_value'] == 'object' ) {
			foreach( $value as $i => $val ) {
				$value[$i] = get_taxonomy( $val );
			}
		}

		return $value;
	}


	/*
	*  Get Taxonomies
	*
	*  Gets taxonomies without the built in ones, but grabs categories and tags
	*
	*/
	function get_taxonomies() {
		$taxonomies = get_taxonomies( array( '_builtin' => false, 'public' => true ), 'objects' );
		$taxonomies['category'] = get_taxonomy('category');
		$taxonomies['post_tag'] = get_taxonomy('post_tag');
		return $taxonomies;
	}

	/*
	*  Get Taxonomies Array
	*
	*  Gets a slug->label array of taxonomies
	*
	*/
	function taxonomies_array() {
		$taxonomies = $this->get_taxonomies();
		$choices = array( 'all' => __( 'All Taxonomies', 'acf-advanced_taxonomy_selector' ) );
		foreach ( $taxonomies as $slug => $taxonomy ) {
			$choices[$slug] = $taxonomy->label;
		}
		return $choices;
	}

	/*
	*  Get Taxonomies From Selection
	*
	*  Gets only those taxonomies which have been selected
	*
	*/
	function get_taxonomies_from_selection( $field ) {
		if( empty( $field['taxonomies'] ) || in_array( 'all', $field['taxonomies'] ) !== false ) {
			$taxonomies = $this->get_taxonomies();
		}
		else {
			$taxonomies = array();
			foreach( $field['taxonomies'] as $taxonomy_slug ) {
				$taxonomy = get_taxonomy( $taxonomy_slug );
				$taxonomies[$taxonomy->name] = $taxonomy;
			}
		}
		return $taxonomies;
	}

}


// create field
new acf_field_advanced_taxonomy_selector();

?>
